_gsocproposals/2020/proposal_ROOTIO.md

```diff
@@ -0,0 +1,49 @@
+---
+title: Pre-conditioners applied to ROOT compression algorithms
+layout: gsoc_proposal
+project: ROOT
+year: 2020
+organization:
+  - UNL
+  - MIR
+---
+
+## Description
+
+Pre-conditioning is an algorithm that rearranges typed, binary data to improve compression. More information can be found here: [1],[2]
+
+The idea of the project is that different filter or transformation schemes that could be applied to ROOT data, which are stored in "baskets". 
+
+In ROOT, the serialization of variable-sized data (containing C-style arrays) produces two internal arrays: one contains the branch data for each event while the other contains the byte offset of each event in the branch data. LZ4 compression algorithm achieves its performance by looking for byte-aligned patterns (as opposed to ZLIB compresison algorithm, which works on individual bits) and lacks the Huffman encoding pass, this results in the offset array sequence being effectively incompressible using LZ4.
+
+Some time ago, Bitshuffle [3] was demonstrated as a possible pre-conditioner for ROOT data with LZ4 for lossless compression [4]. To improve the performance of LZ4 in this case, we investigated the combination of LZ4 with various “pre-conditioners”. Pre-conditioners transform the sequence of input bytes according to a simple, deterministic algorithm prior to applying the compression algorithm. The two algorithms investigated, inspired by the Blosc library [2], are Shuffle and BitShuffle [3]. Both preconditioners rearrange the input array’s bytes by reading through the array using fixed strides. The resulting output of the preconditioner often contains long runs of repeated bytes, improving the compression ratio for LZ4. One of the issues exposed was that it is difficult for ROOT to compress its buffers now due to its 9-byte header.
```

Comment:
"reading through the array" -> "reading through the data" to avoid repetition

Reply:
"long runs" -> "long sequences"?

Reply:
Done! Thanks a lot for the feedback!